### PR TITLE
Prevent ChangeToIenumerable crash on properties.

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes https://github.com/dotnet/roslyn/issues/7087

### CODE REVIEW

File: src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs [hunk 1]
```diff
@@ -50,35 +53,41 @@ public override ImmutableArray<string> FixableDiagnosticIds
...
-            var typeInfo = model.GetTypeInfo(methodDeclarationNode.ReturnType, cancellationToken);
-            if (typeInfo.Type == null)
-            {
-                return null;
-            }
+            
 
-            if (typeInfo.Type.InheritsFromOrEquals(ienumerableSymbol))
+            if (type.GetBaseTypesAndThis().AsImmutable().Concat(type.AllInterfaces).Contains(t =>
+                    SymbolEquivalenceComparer.Instance.Equals(t, ienumerableSymbol)))
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs]: Why the change here?

File: src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs [hunk 2]
```diff
@@ -50,35 +53,41 @@ public override ImmutableArray<string> FixableDiagnosticIds
...
+                root = root.ReplaceNode(node, newMethodDeclarationSyntax);
+            }
+
+            var oldAccessor = (node?.Parent?.Parent as PropertyDeclarationSyntax);
+            if (oldAccessor != null)
+            {
+                root = root.ReplaceNode(oldAccessor, oldAccessor.WithType(newReturnType));
+            }
+
             var newDocument = document.WithSyntaxRoot(root);
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs]: If neither the method declaration case nor the property declaration case are hit, should we still return a CodeAction that does nothing?

File: src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs [hunk 3]
```diff
@@ -32,11 +33,13 @@ public override ImmutableArray<string> FixableDiagnosticIds
         {
             var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             var methodSymbol = model.GetDeclaredSymbol(node, cancellationToken) as IMethodSymbol;
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs]: Probably worth a comment that methodSymbol can be either a regular method or an accessor and to be careful about syntax.

File: src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs [hunk 4]
```diff
@@ -50,35 +53,41 @@ public override ImmutableArray<string> FixableDiagnosticIds
...
             var newReturnType = ienumerableGenericSymbol.GenerateTypeSyntax();
-            var newMethodDeclarationNode = methodDeclarationNode.WithReturnType(newReturnType);
-            root = root.ReplaceNode(methodDeclarationNode, newMethodDeclarationNode);
+            var newMethodDeclarationSyntax = (node as MethodDeclarationSyntax)?.WithReturnType(newReturnType);
+            if (newMethodDeclarationSyntax != null)
+            {
+                root = root.ReplaceNode(node, newMethodDeclarationSyntax);
+            }
+
+            var oldAccessor = (node?.Parent?.Parent as PropertyDeclarationSyntax);
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs]: What about other members?  Indexers?  Operators?  Constructors/Destructors?  etc?

File: src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs (hunk 1 above)
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs]: InheritsFromOrEquals does not check interfaces, only abstract classes, I can add a helper method to make this more succinct.

Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/Iterator/CSharpChangeToIEnumerableCodeFixProvider.cs]: It would seem that that is a separate fix worthy of some tests that fail without the change, then.

**Comment by heejaechang** [comment]
@jmarolf is this test failure specific to your PR?

**Comment by jmarolf** [comment]
@heejaechang I didn't touch tagging, filed an issue

**Comment by leppie** [comment]
Happened to me a while back too, retest fixed it.

**Comment by jmarolf** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/1637/
// Retest reason: flaky test
